drivers/crypto/crypto_rts5912_sha.c

```diff
@@ -0,0 +1,323 @@
+		return ret_val;
+	}
+	k_mutex_lock(&(rts5912_sha256_ctx->crypto_rts5912_in_use), K_FOREVER);
+
+	volatile uint32_t *output_buf = (volatile uint32_t *)output;
+
+	for (uint32_t i = 0; i < 7; i++) {
+		output_buf[i] = sys_cpu_to_be32(rts5912_sha256_ctx->state[i]);
+	}
+	if (rts5912_sha256_ctx->is224 == 0) {
+		output_buf[7] = sys_cpu_to_be32(rts5912_sha256_ctx->state[7]);
+	}
+	k_mutex_unlock(&(rts5912_sha256_ctx->crypto_rts5912_in_use));
+	return ret_val;
+}
+
+static int rts5912_sha256_handler(struct hash_ctx *ctx, struct hash_pkt *pkt, bool finish)
+{
+	struct rts5912_sha256_context *rts5912_sha256_ctx =
+		(struct rts5912_sha256_context *)(ctx->device)->data;
+	int ret_val = 0;
+
+	k_mutex_lock(&(rts5912_sha256_ctx->crypto_rts5912_in_use), K_FOREVER);
+	memcpy(rts5912_sha256_ctx->sha2_data_in_sram, pkt->in_buf, pkt->in_len);
+	k_mutex_unlock(&(rts5912_sha256_ctx->crypto_rts5912_in_use));
+	ret_val = rts5912_sha256_update(ctx->device, rts5912_sha256_ctx->sha2_data_in_sram,
+					pkt->in_len);
+
+	if (ret_val) {
+		return ret_val;
+	} else if (finish) {
+		ret_val = rts5912_sha256_finish(ctx->device, pkt->out_buf);
+	}
+
+	return ret_val;
+}
+
+static int rts5912_hash_begin_session(const struct device *dev, struct hash_ctx *ctx,
+				      enum hash_algo algo)
+{
+	struct rts5912_sha256_context *rts5912_sha256_ctx = dev->data;
+
+	if (rts5912_sha256_ctx->in_use == TRUE) {
+		printk("Crypto driver is busy!");
+		return -EBUSY;
+	}
+
+	rts5912_sha256_ctx->in_use = TRUE;
+
+	switch (algo) {
+	case CRYPTO_HASH_ALGO_SHA224:
+	case CRYPTO_HASH_ALGO_SHA256:
+		k_mutex_lock(&(rts5912_sha256_ctx->crypto_rts5912_in_use), K_FOREVER);
+		memset((void *)(rts5912_sha256_ctx), 0x0, sizeof(struct rts5912_sha256_context));
+		rts5912_sha256_ctx->is224 = (algo == CRYPTO_HASH_ALGO_SHA224);
+		k_mutex_unlock(&(rts5912_sha256_ctx->crypto_rts5912_in_use));
+		ctx->hash_hndlr = rts5912_sha256_handler;
+		rts5912_sha256_start(dev);
+		break;
+	default:
+		return -EINVAL;
+	}
+
+	return 0;
+}
+
+static int rts5912_hash_free_session(const struct device *dev, struct hash_ctx *ctx)
+{
+	struct rts5912_sha256_context *rts5912_sha256_ctx = dev->data;
+
+	rts5912_sha256_ctx->in_use = FALSE;
+
+	return 0;
+}
+
+static inline int rts5912_query_hw_caps(const struct device *dev)
+{
+	return (CAP_SEPARATE_IO_BUFS | CAP_SYNC_OPS);
+}
+
+static int rts5912_sha_init(const struct device *dev)
+{
+	struct rts5912_sha256_context *rts5912_sha256_ctx = dev->data;
+	uint8_t init_buf[32] = {0};
+	struct hash_pkt pkt = {
+		.in_buf = init_buf,
+		.out_buf = init_buf,
+		.in_len = 32,
+		.ctx = NULL,
+	};
+
+	k_mutex_init(&(rts5912_sha256_ctx->crypto_rts5912_in_use));
+	rts5912_sha256_ctx->in_use = FALSE;
+
+	/* For Realtek crypto driver, it has to be inited by running one time. */
+	rts5912_hash_begin_session(dev, pkt.ctx, CRYPTO_HASH_ALGO_SHA256);
+	rts5912_sha256_update(dev, pkt.in_buf, pkt.in_len);
+	rts5912_sha256_finish(dev, pkt.out_buf);
+	rts5912_hash_free_session(dev, pkt.ctx);
```

Comment:
All of these functions can potentially fail (especially `rts5912_sha256_update` and `rts5912_sha256_finish`) so it would be nice to have `rts5912_sha_init` failing if one of these calls fail.